### PR TITLE
Add $file_modified_time column in Iceberg

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Add `$file_modified_time` column in Iceberg

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Add support for hidden `$file_modified_time` columns. ({issue}`13082`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 1]
```diff
@@ -3954,6 +3957,38 @@ public void testPathHiddenColumn()
...
+            assertThat(fileModifiedTime).isBetween(beforeTime, afterTime);
+        }
+    }
+
+    @Test
+    public void testDeleteWithFileModifiedTimeColumn()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_delete_with_file_modified_time_", "(key int)")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (1)", 1);
+            sleepUninterruptibly(100, MILLISECONDS);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:3983]: `1` millisecond should be enough for sleeping.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 2]
```diff
@@ -3954,6 +3957,38 @@ public void testPathHiddenColumn()
...
+            ZonedDateTime afterTime = (ZonedDateTime) computeScalar("SELECT current_timestamp(3)");
+            assertThat(fileModifiedTime).isBetween(beforeTime, afterTime);
+        }
+    }
+
+    @Test
+    public void testDeleteWithFileModifiedTimeColumn()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_delete_with_file_modified_time_", "(key int)")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (1)", 1);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:3982]: `(1)` -> `1` 

The paranthesis are not actually needed in this case.

Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:3982]: I knew, but I believe most test cases use paranthesis in INSERT and there's no such consensuns.

**Review by findinpath — APPROVED** [review]

File: docs/src/main/sphinx/connector/iceberg.rst [hunk 3]
```diff
@@ -622,6 +624,12 @@ Retrieve all records that belong to a specific file using ``"$path"`` filter::
     FROM iceberg.web.page_views
     WHERE "$path" = '/usr/iceberg/table/web.page_views/data/file_01.parquet'
 
+Retrieve all records that belong to a specific file using ``"$file_modified_time"`` filter::
```
Comment (phd3) [inline docs/src/main/sphinx/connector/iceberg.rst:627]: IMO an example with inequality based filtering would be more practical, but nbd

**Review by phd3 — COMMENTED** [review]
thanks @ebyhr, could you please help understand the purpose of this change? If someone wants to look at records at a particular point of time, can't they just use AS OF syntax? 
If the goal is to expose low-level file-status info for every record for debugging, are we planning to also add other stuff in `FileStatus`?

**Comment by ebyhr** [comment]
@phd3 The main purpose is using additional condition in `OPTIMIZE` procedure (requires #13012 though). Another purpose is for debugging as you mentioned, but it's not primary.

**Review by phd3 — APPROVED** [review]
If we're specifically adding this for OPTIMIZE, would it make more sense to provide as an executeProperty rather than an explicit virtual column? however, this seems reasonable from the consistency perspective with delta and hive connector.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 4]
```diff
@@ -493,6 +504,9 @@ else if (partitionKeys.containsKey(column.getId())) {
                 else if (column.isPathColumn()) {
                     columnAdaptations.add(ColumnAdaptation.constantColumn(nativeValueToBlock(FILE_PATH.getType(), utf8Slice(path.toString()))));
                 }
+                else if (column.isFileModifiedTimeColumn()) {
+                    columnAdaptations.add(ColumnAdaptation.constantColumn(nativeValueToBlock(FILE_MODIFIED_TIME.getType(), packDateTimeWithZone(fileModifiedTime.orElseThrow(), DateTimeZone.getDefault().getID()))));
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:508]: Should the zone be configurable or should it be UTC? I think in Delta we do UTC

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataColumn.java [hunk 5]
```diff
@@ -22,11 +22,13 @@
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static io.trino.plugin.iceberg.ColumnIdentity.TypeCategory;
 import static io.trino.plugin.iceberg.ColumnIdentity.TypeCategory.PRIMITIVE;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS;
 import static io.trino.spi.type.VarcharType.VARCHAR;
 
 public enum IcebergMetadataColumn
 {
     FILE_PATH(MetadataColumns.FILE_PATH.fieldId(), "$path", VARCHAR, PRIMITIVE),
+    FILE_MODIFIED_TIME(Integer.MAX_VALUE - 1001, "$file_modified_time", TIMESTAMP_TZ_MILLIS, PRIMITIVE),
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadataColumn.java]: Filed https://github.com/apache/iceberg/issues/5240

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java (hunk 4 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:508]: While it's exactly UTC in Delta, Hive connector uses `DateTimeZone.getDefault()`. 

@findepi Do you have any opinion?

**Comment by ebyhr** [comment]
Just rebased on upstream to resolve conflicts.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java (hunk 4 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java]: Thanks for sharing the context. Changed to UTC.

**Review by alexjo2144 — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 6]
```diff
@@ -297,6 +303,17 @@ public ConnectorPageSource createPageSource(
             fileSize = fileIoProvider.createFileIo(hdfsContext, session.getQueryId())
                     .newInputFile(split.getPath()).getLength();
         }
+        OptionalLong fileModifiedTime = OptionalLong.empty();
+        if (requiredColumns.stream().anyMatch(IcebergColumnHandle::isFileModifiedTimeColumn)) {
+            try {
+                FileStatus fileStatus = hdfsEnvironment.doAs(session.getIdentity(),
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:309]: nit: move `session.getIdentity()` to the next line.

**Review by findinpath — APPROVED** [review]

**Pull request merged**